### PR TITLE
Support IMX8ULP

### PR DESCRIPTION
**Pull request by TangleZ**

### CODE REVIEW

**Comment by TangleZ** [comment]
Have some files license and the author is not correct, will update it later.

File: src/include/sof/lib/dai.h [hunk 1]
```diff
@@ -340,7 +342,6 @@ static inline int dai_trigger(struct dai *dai, int cmd, int direction)
 {
 	int ret = dai->drv->ops.trigger(dai, cmd, direction);
 
-
 	return ret;
```
Comment (dbaluta) [inline src/include/sof/lib/dai.h]: Stray new line here.

File: src/include/sof/lib/dai.h [hunk 2]
```diff
@@ -329,7 +332,6 @@ static inline int dai_set_config(struct dai *dai,
 {
 	int ret = dai->drv->ops.set_config(dai, config);
 
-
```
Comment (dbaluta) [inline src/include/sof/lib/dai.h]: stray new line here.

File: src/include/sof/lib/dai.h [hunk 3]
```diff
@@ -362,7 +362,6 @@ static inline int dai_pm_context_restore(struct dai *dai)
 {
 	int ret = dai->drv->ops.pm_context_restore(dai);
 
-
 	return ret;
```
Comment (dbaluta) [inline src/include/sof/lib/dai.h]: same here. Please check this commit. You have multiple unerlated new lines removals. Do not add them in this commit.

File: src/audio/dai.c [hunk 4]
```diff
@@ -816,6 +816,7 @@ static void dai_config_reset(struct comp_dev *dev)
 static int dai_prepare(struct comp_dev *dev)
```
Comment (dbaluta) [inline src/audio/dai.c:816]: Please explain WHY this patch is needed. Your commit message says WHAT the patch does. While this is also useful the commit message should first explain WHY the commit is needed and then add details about it.

File: src/include/sof/drivers/mu.h [hunk 5]
```diff
@@ -13,27 +13,61 @@
...
+
 /* Status Register */
-#define IMX_MU_xSR		0x20
-#define IMX_MU_xSR_GIPn(x)	BIT(28 + (3 - (x)))
-#define IMX_MU_xSR_RFn(x)	BIT(24 + (3 - (x)))
-#define IMX_MU_xSR_TEn(x)	BIT(20 + (3 - (x)))
+#define IMX_MU_xSR(type, index)        (type & IMX_MU_V2 ? index : 0x20)
+#define IMX_MU_xSR_GIPn(type, x) (type & IMX_MU_V2 ? BIT(x) : BIT(28 + (3 - (x))))
+#define IMX_MU_xSR_RFn(type, x) (type & IMX_MU_V2 ? BIT(x) : BIT(24 + (3 - (x))))
+#define IMX_MU_xSR_TEn(type, x) (type & IMX_MU_V2 ? BIT(x) : BIT(20 + (3 - (x))))
```
Comment (dbaluta) [inline src/include/sof/drivers/mu.h]: Add here type == IMX_MU_V2 instead of type & IMX_MU_V2

File: src/include/sof/drivers/sai.h [hunk 6]
```diff
@@ -247,15 +251,28 @@
  * configured for an internal bit clock.
  * The division value is (DIV + 1) * 2.
  */
+#ifndef CONFIG_IMX8ULP
 #define SAI_CLOCK_DIV		0x7
+#else
+#define SAI_CLOCK_DIV		0x5
+#endif
 #define SAI_TDM_SLOTS		2
```
Comment (dbaluta) [inline src/include/sof/drivers/sai.h]: Why is this needed? there is no explanation in the commit message.

File: src/drivers/imx/sai.c [hunk 7]
```diff
@@ -31,6 +31,7 @@ static void sai_start(struct dai *dai, int direction)
 {
 	dai_info(dai, "SAI: sai_start");
```
Comment (dbaluta) [inline src/drivers/imx/sai.c:32]: Explain what is src id? Maybe create a separate commit introducing src id.

File: src/drivers/imx/interrupt-generic.c [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+#include <sof/spinlock.h>
+#include <errno.h>
+#include <inttypes.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+
+/* fa00558c-d653-4851-a03a-b21f125a9524 */
+DECLARE_SOF_UUID("noirq-imx", noirq_imx_uuid, 0xfa00558c, 0xd653, 0x4851,
+		 0xa0, 0x3a, 0xb2, 0x1f, 0x12, 0x5a, 0x95, 0x24);
```
Comment (dbaluta) [inline src/drivers/imx/interrupt-generic.c:23]: This should be named generic-irq-imx. And thus noirq_imx_uuid should be read as generic_irq_imx_uuid.

File: src/drivers/imx/sai.c [hunk 9]
```diff
@@ -425,6 +430,7 @@ const struct dai_driver sai_driver = {
 		.probe			= sai_probe,
 		.get_handshake		= sai_get_handshake,
 		.get_fifo		= sai_get_fifo,
+		.get_srcid		= sai_get_srcid,
 		.get_hw_params		= sai_get_hw_params,
 	},
```
Comment (dbaluta) [inline src/drivers/imx/sai.c:453]: This patch will no compile. Please add the patches in an order that will not break bisection.

File: src/include/sof/drivers/sai.h [hunk 10]
```diff
@@ -239,6 +241,8 @@
 
 #ifdef CONFIG_IMX8M
 #define SAI_FIFO_WORD_SIZE	128
+#elif defined CONFIG_IMX8ULP
+#define SAI_FIFO_WORD_SIZE	16
 #else
 #define SAI_FIFO_WORD_SIZE	64
 #endif
```
Comment (dbaluta) [inline src/include/sof/drivers/sai.h:248]: add the changes for SAI_OFS and FIFO_WORD_SIZE in a separate patch.

File: src/drivers/imx/sai.c [hunk 11]
```diff
@@ -424,7 +424,11 @@ static int sai_get_hw_params(struct dai *dai,
 
 	/* SAI only currently supports these parameters */
 	params->rate = sai->params.fsync_rate;
+#ifdef CONFIG_IMX8ULP
+	params->channels = 1;
+#else
 	params->channels = 2;
+#endif
 	params->buffer_fmt = 0;
```
Comment (dbaluta) [inline src/drivers/imx/sai.c:432]: I'm fine with this for now. In the future the number of channels should be read from sai->params.

File: src/include/sof/drivers/sai.h [hunk 12]
```diff
@@ -251,8 +251,14 @@
  * configured for an internal bit clock.
  * The division value is (DIV + 1) * 2.
  */
+#ifndef CONFIG_IMX8ULP
 #define SAI_CLOCK_DIV		0x7
 #define SAI_TDM_SLOTS		2
```
Comment (dbaluta) [inline src/include/sof/drivers/sai.h:256]: put the condition the other way around #ifdef CONFIG_IMX8ULP. I'm fine with this for now, but in the future we should make

File: src/include/sof/lib/dai.h [hunk 13]
```diff
@@ -417,6 +420,18 @@ static inline int dai_get_fifo(struct dai *dai, int direction,
...
+			       int stream_id)
+{
+	int ret = dai->drv->ops.get_srcid(dai, direction, stream_id);
+
+
+	return ret;
+}
+
 /**
  * \brief Digital Audio interface Probe
```
Comment (dbaluta) [inline src/include/sof/lib/dai.h:436]: This patch should be much earlier in the series so that we avoid breaking the bisection. E.g after applying each patch the code should still compile.

File: src/include/sof/drivers/mu.h [hunk 14]
```diff
@@ -13,27 +13,61 @@
...
+enum imx_mu_xsr {
+	IMX_MU_SR = 0x0c,
+	IMX_MU_GSR = 0x118,
+	IMX_MU_TSR = 0x124,
+	IMX_MU_RSR = 0x12c,
+};
+
+#ifdef CONFIG_IMX8ULP
+/* Transmit Register */
+#define IMX_MU_xTRn(x)         (0x200 + 4 * (x))
```
Comment (iuliana-prodan) [inline src/include/sof/drivers/mu.h]: Are these registers, `IMX_MU_xTRn`, `IMX_MU_xRRn`, used? 
I couldn't find them anywhere other than:
```
/sof$ grep -rnw IMX_MU_xTRn
tools/fuzzer/platform/mu.h:19:#define IMX_MU_xTRn(x)          (0x00 + 4 * (x))
src/include/sof/drivers/mu.h:17:#define IMX_MU_xTRn(x)		(0x00 + 4 * (x))
```

If they are not used, maybe we should remove them. Add them when need it.

File: src/include/sof/drivers/mu.h [hunk 15]
```diff
@@ -13,27 +13,61 @@
...
 #define IMX_MU_xRRn(x)		(0x10 + 4 * (x))
+#endif
+
 /* Status Register */
-#define IMX_MU_xSR		0x20
-#define IMX_MU_xSR_GIPn(x)	BIT(28 + (3 - (x)))
-#define IMX_MU_xSR_RFn(x)	BIT(24 + (3 - (x)))
-#define IMX_MU_xSR_TEn(x)	BIT(20 + (3 - (x)))
+#define IMX_MU_xSR(type, index)		(type == IMX_MU_V2 ? index : 0x20)
+#define IMX_MU_xSR_GIPn(type, x)	(type == IMX_MU_V2 ? BIT(x) : BIT(28 + (3 - (x))))
```
Comment (iuliana-prodan) [inline src/include/sof/drivers/mu.h]: Can you, please, align this?

File: src/include/sof/drivers/mu.h [hunk 16]
```diff
@@ -13,27 +13,61 @@
...
 /* Receive Register */
 #define IMX_MU_xRRn(x)		(0x10 + 4 * (x))
+#endif
+
 /* Status Register */
-#define IMX_MU_xSR		0x20
-#define IMX_MU_xSR_GIPn(x)	BIT(28 + (3 - (x)))
-#define IMX_MU_xSR_RFn(x)	BIT(24 + (3 - (x)))
-#define IMX_MU_xSR_TEn(x)	BIT(20 + (3 - (x)))
+#define IMX_MU_xSR(type, index)		(type == IMX_MU_V2 ? index : 0x20)
```
Comment (iuliana-prodan) [inline src/include/sof/drivers/mu.h]: Why do you have an `index` here? Shouldn't be a constant like 0x20?
Something like:
`#define IMX_MU_xSR(type, index) (type == IMX_MU_V2 ? const_val_for_version_2: 0x20)`

File: src/include/sof/drivers/mu.h [hunk 17]
```diff
@@ -13,27 +13,61 @@
...
+#define IMX_MU_xSR_GIPn(type, x)	(type == IMX_MU_V2 ? BIT(x) : BIT(28 + (3 - (x))))
+#define IMX_MU_xSR_RFn(type, x)		(type == IMX_MU_V2 ? BIT(x) : BIT(24 + (3 - (x))))
+#define IMX_MU_xSR_TEn(type, x)		(type == IMX_MU_V2 ? BIT(x) : BIT(20 + (3 - (x))))
 #define IMX_MU_xSR_BRDIP	BIT(9)
 
 /* Control Register */
-#define IMX_MU_xCR		0x24
+#define IMX_MU_xCR(type, index)		(type == IMX_MU_V2 ? index : 0x24)
```
Comment (iuliana-prodan) [inline src/include/sof/drivers/mu.h]: Same as above -why do you need the `index`?
If this is not necessary you can also remove the `idx` argument from `imx_mu_xcr_rmw()`, `imx_mu_xsr_rmw()` functions.

File: src/include/sof/drivers/mu.h [hunk 18]
```diff
@@ -45,26 +79,26 @@ static inline void imx_mu_write(uint32_t val, uint32_t reg)
 	*((volatile uint32_t*)(MU_BASE + reg)) = val;
 }
 
-static inline uint32_t imx_mu_xcr_rmw(uint32_t set, uint32_t clr)
+static inline uint32_t imx_mu_xcr_rmw(int type, int idx, uint32_t set, uint32_t clr)
```
Comment (iuliana-prodan) [inline src/include/sof/drivers/mu.h:72]: Please check if `idx` is really necessary.

File: src/drivers/imx/sai.c [hunk 19]
```diff
@@ -88,9 +89,16 @@ static void sai_start(struct dai *dai, int direction)
 			REG_SAI_MCTL_MCLK_EN);
 #endif
 
+	chan_idx = BIT(0);
```
Comment (iuliana-prodan) [inline src/drivers/imx/sai.c:92]: What is `chan_idx`? 
Can you give us some context or maybe point us to the documentation.

File: src/include/sof/drivers/sai.h [hunk 20]
```diff
@@ -18,6 +18,8 @@
 
 #ifdef CONFIG_IMX8M
 #define SAI_OFS		8
+#elif CONFIG_IMX8ULP
```
Comment (iuliana-prodan) [inline src/include/sof/drivers/sai.h:21]: This can be written as:
```
#ifdef CONFIG_IMX8M || CONFIG_IMX8ULP
#define SAI_OFS		8
#else
#define SAI_OFS		0
#endif
```

File: src/include/sof/drivers/sai.h [hunk 21]
```diff
@@ -251,8 +251,14 @@
  * configured for an internal bit clock.
  * The division value is (DIV + 1) * 2.
  */
+#ifndef CONFIG_IMX8ULP
 #define SAI_CLOCK_DIV		0x7
 #define SAI_TDM_SLOTS		2
+#else
+/* expect get frame clk is 16kHz, mono channel and 16bit word on 8ulp */
```
Comment (iuliana-prodan) [inline src/include/sof/drivers/sai.h:258]: s/expect get frame clk is 16kHz, mono channel and 16bit word on 8ulp/ On 8ULP expected frame clk is 16kHz, mono channel and 16 bit word.

File: src/include/sof/lib/dai.h [hunk 22]
```diff
@@ -417,6 +420,18 @@ static inline int dai_get_fifo(struct dai *dai, int direction,
...
 }
 
+/**
+ * \brief Get Digital Audio dma srcid
+ */
+static inline int dai_get_srcid(struct dai *dai, int direction,
+			       int stream_id)
+{
+	int ret = dai->drv->ops.get_srcid(dai, direction, stream_id);
+
```
Comment (iuliana-prodan) [inline src/include/sof/lib/dai.h:430]: Too many newlines.
Also, I believe you can return directly `dai->drv->ops.get_srcid(dai, direction, stream_id);`

File: src/include/sof/lib/dai.h (hunk 13 above)
Comment (iuliana-prodan) [reply src/include/sof/lib/dai.h:436]: Yes, I agree with Daniel.
Definitely must be added earlier, at least before commit no. 2 - "drivers: imx: sai: Add new sai ops: getsrcid"

File: src/include/sof/platform.h [hunk 23]
```diff
@@ -47,6 +47,24 @@ int platform_init(struct sof *sof);
 
 /** @}*/
 
+#define HOST_TO_LOCAL 0
```
Comment (iuliana-prodan) [inline src/include/sof/platform.h:50]: This is i.MX specific, even more is 8ULP specific.
Can we add this in a file imx specific?

File: src/include/sof/lib/dma.h [hunk 24]
```diff
@@ -130,6 +130,7 @@ struct dma_sg_config {
 	uint32_t dest_dev;
 	uint32_t cyclic;			/* circular buffer */
 	uint64_t period;
+	int srcid;
```
Comment (iuliana-prodan) [inline src/include/sof/lib/dma.h:133]: Why do you add `srcid`? Where and for what do you use it?

I see is used in `edma.c` (line 338, above), but instead of 
```
if (srcid) {
...
} 
```
you can use an `#ifdef CONFIG_IMX8ULP`

File: src/platform/imx8ulp/include/arch/xtensa/config/core-isa.h [hunk 25]
```diff
@@ -0,0 +1,673 @@
...
+#define XCHAL_HAVE_HIGHPRI_INTERRUPTS	1	/* med/high-pri. interrupts */
+#define XCHAL_HAVE_NMI			1	/* non-maskable interrupt */
+#define XCHAL_HAVE_CCOUNT		1	/* CCOUNT reg. (timer option) */
+#define XCHAL_NUM_TIMERS		2	/* number of CCOMPAREn regs */
+#define XCHAL_NUM_INTERRUPTS		32	/* number of interrupts */
+#define XCHAL_NUM_INTERRUPTS_LOG2	5	/* ceil(log2(NUM_INTERRUPTS)) */
+#define XCHAL_NUM_EXTINTERRUPTS		27	/* num of external interrupts */
+#define XCHAL_NUM_INTLEVELS		4	/* number of interrupt levels
+						   (not including level zero) */
+#define XCHAL_EXCM_LEVEL		2	/* level masked by PS.EXCM */
```
Comment (iuliana-prodan) [inline src/platform/imx8ulp/include/arch/xtensa/config/core-isa.h:355]: Where did you take this values from?

**Review by iuliana-prodan — CHANGES_REQUESTED** [review]

File: src/include/sof/drivers/mu.h (hunk 16 above)
Comment (TangleZ) [reply src/include/sof/drivers/mu.h]: It's not a constant, according to the code, it will be different register.

File: src/include/sof/drivers/mu.h (hunk 17 above)
Comment (TangleZ) [reply src/include/sof/drivers/mu.h]: The same reason as before.

File: src/include/sof/platform.h (hunk 23 above)
Comment (TangleZ) [reply src/include/sof/platform.h:50]: It should be common code, if we add new platform with need to convert address then this part code will be used. If we add it 
in 8ulp specific, and then we need make it be common when add new platform.

File: src/include/sof/lib/dma.h (hunk 24 above)
Comment (TangleZ) [reply src/include/sof/lib/dma.h:133]: The reason is srcid may be a common resource, when edma really need. And this won't block existing platforms.

File: src/platform/imx8ulp/include/arch/xtensa/config/core-isa.h (hunk 25 above)
Comment (TangleZ) [reply src/platform/imx8ulp/include/arch/xtensa/config/core-isa.h:355]: This file from ./crosstool-ng/.build/xtensa-imx8ulp-elf/build/overlay/newlib/newlib/libc/sys/xtensa/include/xtensa/config/core-isa.h

File: src/include/sof/drivers/mu.h (hunk 18 above)
Comment (TangleZ) [reply src/include/sof/drivers/mu.h:72]: Yes. I'm sure this parameter is needed.

File: src/include/sof/platform.h (hunk 23 above)
Comment (iuliana-prodan) [reply src/include/sof/platform.h:50]: For now is only for 8ulp, so add it just for 8ulp, let's not change the core code if is not needed there.
IF, at some point, will be needed for something else we can make it common, but now this has nothing to do with other imx targets or Intel's.

File: src/include/sof/lib/dma.h (hunk 24 above)
Comment (iuliana-prodan) [reply src/include/sof/lib/dma.h:133]: Commom resource for what? What is used for? 
All I understand right now is that `srcid` is for 8ULP.  But I don't know what it represent - a source id for what?
If is only used to determine it's 8ULP use instead of:
```
if (srcid) {
...
} 
```
you can use an `#ifdef CONFIG_IMX8ULP`

File: src/drivers/imx/edma.c [hunk 26]
```diff
@@ -324,6 +334,14 @@ static int edma_setup_tcd(struct dma_chan_data *channel, int16_t soff,
 	if (rc < 0)
 		return rc;
 
+	/* only impact 8ulp platform */
+	if (srcid) {
```
Comment (iuliana-prodan) [inline src/drivers/imx/edma.c:338]: IF only impacts 8ULP I would put this all in an `#ifdef CONFIG_IMX8ULP`, and remove the chan_addr_convert() function from above.
```
#ifdef CONFIG_IMX8ULP
        if (!dma_chan_reg_read(channel, EDMA_CH_MUX))
		dma_chan_reg_write(channel, EDMA_CH_MUX, srcid);
	if (channel->direction == DMA_DIR_MEM_TO_DEV)
		convert_addr(LOCAL_TO_HOST, sbase);
	else
		convert_addr(LOCAL_TO_HOST, dbase);
#endif
```
Is not ideal, but to add a variable, `srcid`, just for this I believe is worst.

Also, maybe we can add the `convert_addr()` in a` src/driver/imx/utils.c` file? 
@dbaluta @TangleZ what do you think?

File: tools/topology/CMakeLists.txt [hunk 27]
```diff
@@ -136,6 +136,7 @@ set(TPLGS
 	"sof-jsl-da7219\;sof-jsl-da7219\;-DPLATFORM=jsl"
 	"sof-jsl-da7219\;sof-jsl-da7219-mx98360a\;-DPLATFORM=jsl-dedede"
 	"sof-imx8mp-wm8960\;sof-imx8mp-wm8960"
+	"sof-imx8ulp-btsco\;sof-imx8ulp-btsco"
```
Comment (iuliana-prodan) [inline tools/topology/CMakeLists.txt:139]: Why did you added -btsco? 
Is this related with the Bluetooth SCO Audio - https://man.netbsd.org/btsco.4

**Review by iuliana-prodan — CHANGES_REQUESTED** [review]

File: src/include/sof/platform.h (hunk 23 above)
Comment (TangleZ) [reply src/include/sof/platform.h:50]: Sure.

File: tools/topology/CMakeLists.txt (hunk 27 above)
Comment (TangleZ) [reply tools/topology/CMakeLists.txt:139]: Yes. sai will connect to Bluetooth.

File: src/audio/dai.c [hunk 28]
```diff
@@ -838,6 +839,7 @@ static int dai_prepare(struct comp_dev *dev)
 		comp_set_state(dev, COMP_TRIGGER_RESET);
 		return -EINVAL;
 	}
+	dd->chan->srcid = dai_get_srcid(dd->dai, dai->direction, dd->stream_id);
```
Comment (dbaluta) [inline src/audio/dai.c:843]: We need to think about this. For non-8ulp platforms this function will crash. At least return 0 if src_id is not implemented.

File: src/include/sof/lib/dai.h [hunk 29]
```diff
@@ -417,6 +420,18 @@ static inline int dai_get_fifo(struct dai *dai, int direction,
 	return ret;
 }
 
+/**
+ * \brief Get Digital Audio dma srcid
+ */
+static inline int dai_get_srcid(struct dai *dai, int direction,
+			       int stream_id)
+{
+	int ret = dai->drv->ops.get_srcid(dai, direction, stream_id);
```
Comment (dbaluta) [inline src/include/sof/lib/dai.h:429]: return 0 here if ops.get_srcid is NULL.

File: src/include/sof/lib/dma.h (hunk 24 above)
Comment (TangleZ) [reply src/include/sof/lib/dma.h:133]: The name srcid is a copy name from edma driver, I think its meaning should be a source id, it represent dmamux number.

File: src/include/sof/lib/dai.h (hunk 29 above)
Comment (TangleZ) [reply src/include/sof/lib/dai.h:429]: Yes. It should be 0 if the platform is not 8ULP. Then edma won't set srcid.

File: src/drivers/imx/edma.c (hunk 26 above)
Comment (dbaluta) [reply src/drivers/imx/edma.c:338]: @TangleZ I agree with @iuliana-prodan. We could do this change locally to our imx edma driver and remove the changes in SOF core.
In the future if other drivers will need this we can reintroduce the get_srcid function.

So, as far as I understand srcid is to map between DMA events and DMA chnannels right? 
And this is done via a MUX?

Please explain this in the commit message and go with a solution that follows Iulia's suggestion which doesn't impact SOF core.

File: tools/topology/sof/pipe-volume-playback.m4 [hunk 30]
```diff
@@ -80,7 +80,8 @@ indir(`define', concat(`PIPELINE_PCM_', PIPELINE_ID), Passthrough Playback PCM_I
 # PCM Configuration
 
 #
-PCM_CAPABILITIES(Passthrough Playback PCM_ID, CAPABILITY_FORMAT_NAME(PIPELINE_FORMAT), PCM_MIN_RATE, PCM_MAX_RATE, 2, PIPELINE_CHANNELS, 2, 16, 192, 16384, 65536, 65536)
+PCM_CAPABILITIES(Passthrough Playback PCM_ID, CAPABILITY_FORMAT_NAME(PIPELINE_FORMAT),
+			PCM_MIN_RATE, PCM_MAX_RATE, PIPELINE_CHANNELS, PIPELINE_CHANNELS, 2, 16, 192, 16384, 65536, 65536)
```
Comment (iuliana-prodan) [inline tools/topology/sof/pipe-volume-playback.m4:84]: You removed 2 (as the channels_min) at all, you should change it to 1, not remove it at all.

Comment (TangleZ) [reply tools/topology/sof/pipe-volume-playback.m4:84]: Not is removed it, just make channel_max = channel_min as in other files. Please refer pipe-volume-capture.m4

File: src/drivers/imx/edma.c (hunk 26 above)
Comment (TangleZ) [reply src/drivers/imx/edma.c:338]: @dbaluta @iuliana-prodan Not very clear about this. SRCID is not related to convert addr function, they can't be mixed.
If you think the code of srcid part is changed SOF core, then I can use hardcode to fix it in edma driver.
If you think convert addr part is changed SOF core, then we can do as iuliana-prodan said, create new file named utils.c for it.
